datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -1917,6 +1923,84 @@ impl Filter {
 
         Ok(Self { predicate, input })
     }
+
+    fn is_scalar(&self) -> bool {
+        let schema = self.input.schema();
+
+        let functional_dependencies = self.input.schema().functional_dependencies();
+        let unique_keys = functional_dependencies.iter().filter(|dep| {
+            let nullable = dep.nullable
+                && dep
+                    .source_indices
+                    .iter()
+                    .any(|&source| schema.field(source).is_nullable());
+            !nullable
+                && dep.mode == Dependency::Single
+                && dep.target_indices.len() == schema.fields().len()
+        });
+
+        /// Splits a conjunctive [`Expr`] such as `A AND B AND C` => `[A, B, C]`
+        pub fn split_conjunction(expr: &Expr) -> Vec<&Expr> {
```

Comment:
I believe this is a duplicate of https://docs.rs/datafusion/latest/datafusion/optimizer/utils/fn.split_conjunction.html

Perhaps we could move the function into https://github.com/apache/arrow-datafusion/blob/main/datafusion/expr/src/utils.rs instead so we can avoid the copy/paste?

Reply:
There are quite a few functions in `datafusion-optimizer::utils` which only depend on `LogicalPlan` or `Expr` and are good candidates to move to `datafusion-expr::utils`. I'll file a follow-up PR for that.

Reply:
I'll rebase tomorrow to get rid of this TODO